python-checks/src/main/java/org/sonar/python/checks/cdk/PublicNetworkAccessToCloudResourcesCheck.java

```diff
@@ -51,10 +54,58 @@ protected void registerFqnConsumer() {
     checkFqn("aws_cdk.aws_rds.DatabaseInstance", PublicNetworkAccessToCloudResourcesCheck::checkDatabaseInstance);
 
     checkFqn("aws_cdk.aws_rds.CfnDBInstance", (subscriptionContext, callExpression) ->
-      getArgument(subscriptionContext, callExpression, PUBLICLY_ACCESSIBLE_ARG_NAME).ifPresent(
-        argument -> argument.addIssueIf(isTrue(), MESSAGE)
-      )
-    );
+      getArgument(subscriptionContext, callExpression, PUBLICLY_ACCESSIBLE_ARG_NAME).ifPresent(argument -> {
+        if (argument.hasExpression(isTrue()) && !isSafeNetwork(subscriptionContext, callExpression)) {
+          argument.addIssue(MESSAGE);
+        }
+      }));
+  }
+
+  private static boolean isSafeNetwork(SubscriptionContext subscriptionContext, CallExpression callExpression) {
+    Optional<QualifiedExpression> subnetGroupName = getArgument(subscriptionContext, callExpression, "db_subnet_group_name")
+      .flatMap(expressionFlow -> expressionFlow.getExpression(isQualifiedExpression()))
+      .map(QualifiedExpression.class::cast);
+
+    if (subnetGroupName.isEmpty()) {
+      return false;
+    }
+
+    QualifiedExpression qualifiedExpression = subnetGroupName.get();
+    if (!"ref".equals(qualifiedExpression.name().name())) {
+      return false;
+    }
+
+    ExpressionFlow qualifierFlow = ExpressionFlow.build(subscriptionContext, qualifiedExpression.qualifier());
+    Optional<QualifiedExpression> subnetIds = qualifierFlow.getExpression(isCallExpression().and(isFqn("aws_cdk.aws_rds.CfnDBSubnetGroup")))
+      .map(CallExpression.class::cast)
+      .flatMap(call -> getArgument(subscriptionContext, call, "subnet_ids"))
+      .flatMap(inner -> inner.getExpression(isQualifiedExpression()))
+      .map(QualifiedExpression.class::cast);
+
+    if (subnetIds.isEmpty()) {
+      return false;
+    }
+
+    if (!"subnet_ids".equals(subnetIds.get().name().name())) {
+      return false;
+    }
+
+    ExpressionFlow subnetIdVpcQualifier = ExpressionFlow.build(subscriptionContext, subnetIds.get().qualifier());
+    Optional<CallExpression> vpcSelectSubnetCall = subnetIdVpcQualifier
+      .getExpression(isCallExpression())
+      .map(CallExpression.class::cast);
+
+    if (vpcSelectSubnetCall.isPresent()) {
+      return getArgument(subscriptionContext, vpcSelectSubnetCall.get(), "subnet_type")
+        .map(subnetType -> subnetType.hasExpression(isFqn("aws_cdk.aws_ec2.SubnetType.PUBLIC").negate()))
+        .orElse(false);
```

Comment:
As I understand the ticket we should not raise an issue on this case:
```
rds_subnet_group_unknown_with_subnet_ids = rds.CfnDBSubnetGroup(
    subnet_ids=ec2.SubnetType.PRIVATE_WITH_NAT
)

rds.CfnDBInstance(
    publicly_accessible=True,
    db_subnet_group_name=rds_subnet_group_unknown_with_subnet_ids.ref
)
```
But we do. So I think is logic is incorrect but due to the Optional magic, we can not see it on the coverage.